### ISSUE TITLE
Hard Linked Files Break on Write

### ISSUE THREAD

**Issue by harryhmtse** [helix-editor/helix#12058]
### Summary

When editting a file, consisting of multiple copies via file system hard links, when Helix writes out the changes to the file.  The hard links are not preserved.  A new file is written out, and broken out of the chain of hard linked files.  Incumbent editors, like vim, nano preserve hard links and all the files in the chain are changed in tandem.   

### Reproduction Steps

1.  Create a File with Text
```
echo "The Big Brown Fox Jumps Over the Lazy Dog" > file.txt
cat file.txt
```
2 . Create Hard LInked Copies of file.txt
```
ln file.txt file-1.txt
ln file.txt file-2.txt
ln file.txt file-3.txt
```
3. Verify number of hard linked files
```
helix:~ # ls -la
total 16
drwx------ 1 root root  94 Nov 12 23:02 .
drwxr-xr-x 1 root root 138 Nov 12 22:59 ..
drwx------ 1 root root   0 Sep 25 19:10 .gnupg
drwxr-xr-x 1 root root   0 Sep 25 19:10 bin
-rw-r--r-- 4 root root  42 Nov 12 23:01 file-1.txt
-rw-r--r-- 4 root root  42 Nov 12 23:01 file-2.txt
-rw-r--r-- 4 root root  42 Nov 12 23:01 file-3.txt
-rw-r--r-- 4 root root  42 Nov 12 23:01 file.txt
```
4. Edit file.txt and make a change in helix
```
hx file.txt
```
5. Verify the file editted is removed from the hard link chain, and the other files remain unchanged.  
```
helix:~ # ls -la
total 16
drwx------ 1 root root 120 Nov 12 23:02 .
drwxr-xr-x 1 root root 138 Nov 12 22:59 ..
drwxr-xr-x 1 root root  10 Nov 12 23:02 .cache
drwxr-xr-x 1 root root  10 Nov 12 23:02 .config
drwx------ 1 root root   0 Sep 25 19:10 .gnupg
drwxr-xr-x 1 root root   0 Sep 25 19:10 bin
-rw-r--r-- 3 root root  42 Nov 12 23:01 file-1.txt
-rw-r--r-- 3 root root  42 Nov 12 23:01 file-2.txt
-rw-r--r-- 3 root root  42 Nov 12 23:01 file-3.txt
-rw-r--r-- 1 root root  44 Nov 12 23:02 file.txt
```

I expected this to happen:

All the files in the hard link chain are preserved and the changes to the file made in Helix are reflected in all copies.   

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2024-11-12T23:36:33.197 globset [DEBUG] glob converted to regex: Glob { glob: "*/Dockerfile.*", re: "(?-u)^.*/Dockerfile\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('D'), Literal('o'), Literal('c'), Literal('k'), Literal('e'), Literal('r'), Literal('f'), Literal('i'), Literal('l'), Literal('e'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.197 globset [DEBUG] glob converted to regex: Glob { glob: "*/dockerfile.*", re: "(?-u)^.*/dockerfile\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('d'), Literal('o'), Literal('c'), Literal('k'), Literal('e'), Literal('r'), Literal('f'), Literal('i'), Literal('l'), Literal('e'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.197 globset [DEBUG] glob converted to regex: Glob { glob: "*/Containerfile.*", re: "(?-u)^.*/Containerfile\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('C'), Literal('o'), Literal('n'), Literal('t'), Literal('a'), Literal('i'), Literal('n'), Literal('e'), Literal('r'), Literal('f'), Literal('i'), Literal('l'), Literal('e'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.197 globset [DEBUG] glob converted to regex: Glob { glob: "*/containerfile.*", re: "(?-u)^.*/containerfile\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('c'), Literal('o'), Literal('n'), Literal('t'), Literal('a'), Literal('i'), Literal('n'), Literal('e'), Literal('r'), Literal('f'), Literal('i'), Literal('l'), Literal('e'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.197 globset [DEBUG] glob converted to regex: Glob { glob: "*/.*ignore", re: "(?-u)^.*/\\..*ignore$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('.'), ZeroOrMore, Literal('i'), Literal('g'), Literal('n'), Literal('o'), Literal('r'), Literal('e')]) }
2024-11-12T23:36:33.197 globset [DEBUG] glob converted to regex: Glob { glob: "*/BUILD.*", re: "(?-u)^.*/BUILD\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('B'), Literal('U'), Literal('I'), Literal('L'), Literal('D'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.198 globset [DEBUG] glob converted to regex: Glob { glob: "*/.env.*", re: "(?-u)^.*/\\.env\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('.'), Literal('e'), Literal('n'), Literal('v'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.198 globset [DEBUG] glob converted to regex: Glob { glob: "*/.envrc.*", re: "(?-u)^.*/\\.envrc\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('.'), Literal('e'), Literal('n'), Literal('v'), Literal('r'), Literal('c'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.198 globset [DEBUG] glob converted to regex: Glob { glob: "*/conf/*/*.{inc,conf}", re: "(?-u)^.*/conf/.*/.*\\.(?:conf|inc)$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('c'), Literal('o'), Literal('n'), Literal('f'), Literal('/'), ZeroOrMore, Literal('/'), ZeroOrMore, Literal('.'), Alternates([Tokens([Literal('c'), Literal('o'), Literal('n'), Literal('f')]), Tokens([Literal('i'), Literal('n'), Literal('c')])])]) }
2024-11-12T23:36:33.198 globset [DEBUG] glob converted to regex: Glob { glob: "*/Jenkinsfile.*", re: "(?-u)^.*/Jenkinsfile\\..*$", opts: GlobOptions { case_insensitive: false, literal_separator: false, backslash_escape: true, empty_alternates: false }, tokens: Tokens([ZeroOrMore, Literal('/'), Literal('J'), Literal('e'), Literal('n'), Literal('k'), Literal('i'), Literal('n'), Literal('s'), Literal('f'), Literal('i'), Literal('l'), Literal('e'), Literal('.'), ZeroOrMore]) }
2024-11-12T23:36:33.198 globset [DEBUG] built glob set; 1 literals, 0 basenames, 0 extensions, 0 prefixes, 129 suffixes, 9 required extensions, 10 regexes
2024-11-12T23:36:33.203 helix_view::clipboard::provider [DEBUG] No native clipboard provider found. Yanking by OSC 52 and pasting will be internal to Helix
2024-11-12T23:36:33.205 helix_vcs [DEBUG] Error {
    context: "failed to open git repo",
    source: Discover(
        NoGitRepository {
            path: "/root",
        },
    ),
}
2024-11-12T23:36:33.205 helix_vcs [DEBUG] failed to open diff base for /root/file.txt
2024-11-12T23:36:33.205 helix_vcs [DEBUG] Error {
    context: "failed to open git repo",
    source: Discover(
        NoGitRepository {
            path: "/root",
        },
    ),
}
2024-11-12T23:36:33.205 helix_vcs [DEBUG] failed to obtain current head name for /root/file.txt
2024-11-12T23:36:33.205 helix_view::editor [DEBUG] editor status: Loaded 1 file.
2024-11-12T23:36:33.357 helix_tui::backend::crossterm [DEBUG] The keyboard enhancement protocol is not supported in this terminal (checked in 150.862394ms)
2024-11-12T23:36:33.357 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:33.454 helix_term::application [DEBUG] received editor event: IdleTimer
2024-11-12T23:36:38.311 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:38.432 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:38.563 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:38.563 helix_term::application [DEBUG] received editor event: IdleTimer
2024-11-12T23:36:38.928 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:39.049 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:39.179 helix_term::application [DEBUG] received editor event: IdleTimer
2024-11-12T23:36:39.183 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:39.305 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2024-11-12T23:36:39.434 helix_term::application [DEBUG] received editor event: IdleTimer
2024-11-12T23:36:40.555 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 1
2024-11-12T23:36:40.806 helix_term::application [DEBUG] received editor event: IdleTimer
2024-11-12T23:36:40.948 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 1
2024-11-12T23:36:41.117 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 1
2024-11-12T23:36:41.199 helix_term::application [DEBUG] received editor event: IdleTimer
2024-11-12T23:36:41.288 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 1
2024-11-12T23:36:41.565 helix_view::document [DEBUG] submitting save of doc 'Some("/root/file.txt")'
2024-11-12T23:36:41.565 helix_term::job [DEBUG] waiting on jobs...
2024-11-12T23:36:41.566 helix_view::document [DEBUG] doc 1 revision updated 0 -> 1
2024-11-12T23:36:41.566 helix_term::commands::typed [DEBUG] quitting...
2024-11-12T23:36:41.566 helix_view::document [DEBUG] id 1 modified - last saved: 1, current: 1
2024-11-12T23:36:41.566 helix_term::job [DEBUG] waiting on jobs...
2024-11-12T23:36:41.566 helix_term::job [DEBUG] waiting on jobs...
2024-11-12T23:36:41.566 helix_lsp::file_event [DEBUG] Received file event for "/root/file.txt"```

</details>


### Platform

Linux, macOS

### Terminal Emulator

Terminal.app, iTerm2, GNOME Console

### Installation Method

GitHub releases, zypper, apt

### Helix Version

helix 24.7 (079f5442)

**Comment by kirawi** [comment]
Duplicate of https://github.com/helix-editor/helix/issues/11279
